### ISSUE TITLE
@theia/task Doesn't run task, creates new terminal

### ISSUE THREAD

**Issue by shortwavedave** [theia-ide/theia#5825]
### Description
Terminal/Run task does not execute the example tasks in the tasks readme. In my zsh shell environment it opens new terminals. To verify it wasn't my environment, I spun up a vm in gcloud with a bash environment. Here are some gifs to demonstrate:

- On my zsh environment:

![theiaTaskIssue](https://user-images.githubusercontent.com/22798574/62189327-b9be8480-b37f-11e9-9b6f-2cc355c2ba49.gif)

- On the gcloud bash environment:

![theiaTaskIssue2](https://user-images.githubusercontent.com/22798574/62190451-70bbff80-b382-11e9-9075-6cf0fc4e2413.gif)

### Reproduction Steps
This is while developing my own extension, so to be sure I started a new "hello world" extension project using the tutorial steps. Then I go into "browser-app", execute:

`yarn add @theia/task`

to add the extension. I also tried running yarn add in my extension directory.

Restart the project and try to run a task.

**OS and Theia version:**
- zsh environment:

> Linux pop-os 5.0.0-21-generic #22+system76-Ubuntu SMP Tue Jul 16 19:57:52 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

-gcloud bash environment:

> Linux 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1+deb9u3 (2019-06-16) x86_64 GNU/Linux

**Diagnostics:** 
There the errors that I see: :

> 0|api      | root WARN Collided keybinding is ignored;  {"command":"outlineView:toggle","keybinding":"ctrlcmd+shift+s"}  collided with  {"command":"file.saveAs","keybinding":"ctrl+shift+s"}
> 0|api      | root WARN Could not register keybinding:
> 0|api      |   {"command":"outlineView:toggle","keybinding":"ctrlcmd+shift+s"}
> 0|api      | Error: "ctrlcmd+shift+s" is in collision with something else [scope:0]
> 0|api      | root WARN Collided keybinding is ignored;  {"command":"passthrough","keybinding":"ctrl+k","context":"terminalActive"}  collided with  {"command":"terminal:clear","keybinding":"ctrlcmd+k","context":"terminalActive"}
> 0|api      | root WARN Could not register keybinding:
> 0|api      |   {"command":"passthrough","keybinding":"ctrl+k","context":"terminalActive"}
> 0|api      | Error: "ctrl+k" is in collision with something else [scope:0]
> 0|api      | root WARN e.onStart is slow, took: 108.82500000298023 ms
> 0|api      | root WARN e.onStart is slow, took: 1282.455000007758 ms
> 0|api      | root ERROR Error(s) reading config file: file:///home/david/georq_workspace/.theia/tasks.json
> 0|api      | root ERROR Couldn't resize terminal -1, because it doesn't exist.
> 0|api      | root ERROR Couldn't resize terminal -1, because it doesn't exist.
> 0|api      | terminal ERROR Couldn't attach - can't find terminal with id: 12 
> 0|api      | terminal ERROR Error attaching to terminal id 12, the terminal is most likely gone. Starting up a new terminal instead.

On the zsh environment it is a little more verbose:

> Started watching: /home/david/src/map_workspace/.theia/tasks.json
> task INFO Created new task, id: 0, process id: 10, OS PID: 14940, context: file:///home/david/src/map_workspace
> terminal ERROR Couldn't attach - can't find terminal with id: 10 
> root ERROR Couldn't resize terminal -1, because it doesn't exist.
> terminal ERROR Error attaching to terminal id 10, the terminal is most likely gone. Starting up a new terminal instead.

So clearly, the task extension cannot find the current terminal and then starts a new terminal, but doesn't run the task either way.

**Comment by akosyakov** [comment]
A new terminal for a task is expected. Although it should be reused for executing the same task again. It is known issue.

From your gifs it looks like taks is actually not properly executed, i.e. output is not there, correct?

**Comment by shortwavedave** [comment]
> From your gifs it looks like taks is actually not properly executed, i.e. output is not there, correct?
> 
right, the output is not there. So I guess if I want to execute a python script programmatically I should just use the process extension.

**Comment by akosyakov** [comment]
> right, the output is not there.

It's a bug.

> So I guess if I want to execute a python script programmatically I should just use the process extension.

If you don't want to show output to the end user, yes. If you want then tasks should be handy.

**Comment by shortwavedave** [comment]
FYI, this is the extension version from yarn.lock

> "@theia/task@^0.9.0":
>   resolved "https://registry.yarnpkg.com/@theia/task/-/task-0.9.0.tgz#751079605c95289ab71c3c43720bcf8634a8e310"

**Comment by vince-fugnitto** [comment]
I believe it's a duplicate of the following issue https://github.com/theia-ide/theia/issues/2961

It looks like the output of short lived tasks are not properly outputted since they complete before a terminal can be created and initialized.

**Comment by akosyakov** [comment]
@vince-fugnitto some is going to look at it? i.e. @elaihau ? fixing such bugs is important for vscode support as well. Or should we mark it as `help wanted`?

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto some is going to look at it? i.e. @elaihau ? fixing such bugs is important for vscode support as well. Or should we mark it as `help wanted`?

I can inform @elaihau about the bug and see if he wants to handle it. He's done some great improvements to the `tasks` system already.

**Comment by eiso** [comment]
I have encountered the same issue. 

To replicate:

```TS
 protected getWorkspaceContext(): string | undefined {
        return this.workspaceService.workspace && this.workspaceService.workspace.uri;
    }

    async launchTask(arg:string): Promise<void> {
        
        const taskConfig: TaskConfiguration = {
            label: 'test runner',
            type: 'shell',
            command: '/path/to/testscript',
            options: {
                cwd: this.getWorkspaceContext()
            },
            _source: this.getWorkspaceContext(),
            args: [
                //arg
            ]
        };

        let taskInfo: TaskInfo;

        try {
            console.log('Executing this task');
            taskInfo = await this.taskServer.run(taskConfig);
            if (taskInfo.terminalId !== undefined) {
                console.log(taskInfo)
                this.logger.debug(`Task created: taskInfo: ${taskInfo}`);
                this.taskService.attach(taskInfo.terminalId, taskInfo.taskId);
            }
            console.log("SUCCESS!")
        } catch (error) {
            const errorStr = `Error launching task: ${error.message}`;
            this.logger.error(errorStr);
            return;
        }
    }
```

Example output:

```
task INFO Created new task, id: 5, process id: 16, OS PID: 49171, context: [object Object]
root INFO Executing this task
root INFO file:///path/to/somewhere
root INFO { taskId: 5,
  ctx: { _isCancelled: false },
  config: 
   { label: 'python runner',
     type: 'shell',
     command: '/path/to/testscript',
     options: { cwd: 'file:///path/to/somewhere', env: [Object] },
     _source: 'file:///path/to/somewhere',
     args: [] },
  terminalId: 16 }
root INFO SUCCESS!
terminal ERROR Couldn't attach - can't find terminal with id: 16 
terminal ERROR Error attaching to terminal id 16, the terminal is most likely gone. Starting up a new terminal instead.
root ERROR Couldn't resize terminal -1, because it doesn't exist.
```

What is most interesting is that the process that gets started when a task is created, has a PID on my system of `+1` of the PID that is logged. So in this case `49171` is running as `/bin/bash -l`. The testscript path provided is never executed.

**Comment by akosyakov** [comment]
@elaihau @RomanNikitenko  Was it resolved by https://github.com/eclipse-theia/theia/pull/6668?

**Comment by elaihau** [comment]
> @elaihau @RomanNikitenko Was it resolved by #6668?

I think not. 

@akosyakov 
I believe this issue and #6668 are unrelated, as this one is caused by "the short running task finishes running before being attached to the terminal".

**Comment by akosyakov** [comment]
@elaihau ok, do you know anyone is looking into https://github.com/eclipse-theia/theia/issues/2961?

**Comment by elaihau** [comment]
It is not an easy change. From what I can remember Sven had a PR for it (not merged). 
@marechal-p also looked at it but he had other priorities so didn't get the chance to invest much time on it.

**Comment by marechal-p** [comment]
It is somewhere in my backlog indeed, but while we are at it, I would appreciate input on what I was thinking about doing:

The core of the issue is that when a process exits/closes, we dispose everything related to it. This is an issue when the execution is fast enough that the frontend cannot request for output quick enough, since it is being discarded as fast as possible.

I was thinking about storing processes and/or output in some "waiting" list, where the output would stay as long as either the frontend comes and "collects" whatever data is required, or wait for some timeout to dispose everything if the frontend didn't care about it.

This would mimic the way the Unix kernel handles processes: When a process dies, it stays into the process table until its parent calls [`waitpid()`](https://linux.die.net/man/2/waitpid). Until the parent process makes that call, the child is listed as "zombie" or "defunct", and will eventually get collected by the `init` process if the parent exits and forgets to call `waitpid`.

@akosyakov @elaihau any opinions on that approach?
